### PR TITLE
Better doc for `jax.numpy.i0`

### PR DESCRIPTION
**Pull request by rajasekharporeddy**
Part of #21461

### CODE REVIEW

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -5973,19 +5973,44 @@ def meshgrid(*xi: ArrayLike, copy: bool = True, sparse: bool = False,
   return output
 
 
-@custom_jvp
-@util.implements(np.i0)
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:5977]: Out of curiosity, why did you move the custom JVP from here to a separate implementation function?

File: jax/_src/numpy/lax_numpy.py [hunk 2]
```diff
@@ -5973,19 +5973,44 @@ def meshgrid(*xi: ArrayLike, copy: bool = True, sparse: bool = False,
...
 @jit
 def i0(x: ArrayLike) -> Array:
+  """Calculate modified Bessel function of first kind, zeroth order.
+
+  JAX implementation of :func:`numpy.i0`.
+
+  Args:
+    x: scalar or array. Specifies the argument of Bessel function.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:5983]: Mention that complex inputs are not supported

File: jax/_src/numpy/lax_numpy.py [hunk 3]
```diff
@@ -5973,19 +5973,44 @@ def meshgrid(*xi: ArrayLike, copy: bool = True, sparse: bool = False,
   return output
 
 
-@custom_jvp
-@util.implements(np.i0)
 @jit
 def i0(x: ArrayLike) -> Array:
+  """Calculate modified Bessel function of first kind, zeroth order.
+
+  JAX implementation of :func:`numpy.i0`.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:5980]: It might be nice to add the mathematical formula here (you can copy it from jax.scipy's `i0` docstring).

File: jax/_src/numpy/lax_numpy.py [hunk 4]
```diff
@@ -5973,19 +5973,44 @@ def meshgrid(*xi: ArrayLike, copy: bool = True, sparse: bool = False,
...
   x_arr, = util.promote_args_inexact("i0", x)
   if not issubdtype(x_arr.dtype, np.floating):
     raise ValueError(f"Unsupported input type to jax.numpy.i0: {_dtype(x)}")
-  x_arr = lax.abs(x_arr)
+  return _i0(x_arr)
+
+
+@custom_jvp
+def _i0(x):
+  x_arr = lax.abs(x)
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:6008]: This is strange variable naming in this new context. Previously it was called `x_arr` because the value was `x` converted to an array. Here maybe call it `abs_x`?

File: jax/_src/numpy/lax_numpy.py [hunk 5]
```diff
@@ -5973,19 +5973,44 @@ def meshgrid(*xi: ArrayLike, copy: bool = True, sparse: bool = False,
...
+  JAX implementation of :func:`numpy.i0`.
+
+  Args:
+    x: scalar or array. Specifies the argument of Bessel function.
+
+  Returns:
+    An array containing the corresponding vlaues of the modified Bessel function
+    of ``x``.
+
+  See also:
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:5989]: You could also mention `jax.scipy.special.i0e`

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/numpy/lax_numpy.py (hunk 1 above)
Comment (rajasekharporeddy) [reply jax/_src/numpy/lax_numpy.py:5977]: The documentation of the API was not updated when I build the docs locally with `custom_jvp` as decorator to `i0`. However when I moved it to a separate implementation function, the document of the API is updated.

Comment (jakevdp) [reply jax/_src/numpy/lax_numpy.py:5977]: Maybe it was a cacheing issue? `custom_jvp` should be fine for docs. For example `relu` is decorated with `custom_jvp` and the docs render without an issue: https://jax.readthedocs.io/en/latest/_autosummary/jax.nn.relu.html

File: jax/_src/numpy/lax_numpy.py (hunk 5 above)
Comment (rajasekharporeddy) [reply jax/_src/numpy/lax_numpy.py:5996]: Done

File: jax/_src/numpy/lax_numpy.py (hunk 4 above)
Comment (rajasekharporeddy) [reply jax/_src/numpy/lax_numpy.py]: Done.

File: jax/_src/numpy/lax_numpy.py (hunk 3 above)
Comment (rajasekharporeddy) [reply jax/_src/numpy/lax_numpy.py:5980]: Done

File: jax/_src/numpy/lax_numpy.py (hunk 2 above)
Comment (rajasekharporeddy) [reply jax/_src/numpy/lax_numpy.py]: Done

**Comment by rajasekharporeddy** [comment]
Fixed the lint_and_typecheck error. Could you please trigger the tests again?

**Review by jakevdp — APPROVED** [review]
Looks great, thanks!

**Comment by jakevdp** [comment]
Can you please squash the changes into a single commit?

**Comment by rajasekharporeddy** [comment]
Thanks! Squashed into single commit.
